chalk-ir/src/visit.rs

```diff
@@ -0,0 +1,421 @@
+//! Traits for visiting bits of IR.
+use std::fmt::Debug;
+
+use crate::{
+    BoundVar, DebruijnIndex, DomainGoal, Goal, InferenceVar, Interner, Lifetime, LifetimeData,
+    PlaceholderIndex, ProgramClause, Ty, TyData, WhereClause,
+};
+
+mod binder_impls;
+mod boring_impls;
+pub mod visitors;
+
+pub use visitors::VisitExt;
+
+pub trait VisitResult: Sized {
+    fn new() -> Self;
+
+    fn return_early(&self) -> bool;
+    fn combine(self, other: Self) -> Self;
+
+    fn and_then(self, op: impl FnOnce() -> Self) -> Self {
```

Comment:
Comment:

Convenience helper for use in writing `Visitor` impls. Returns `self` if `return_early()` is true, but otherwise combines `self` with the result of invoking `op`.

If you have a struct with two fields, `foo` and `bar`, you can thus write code like

```rust,ignore
self.foo.visit_with(visitor, outer_binder)
    .and_then(|| self.bar.visit_with(visitor, outer_binder))
```

and `bar` will only be visited if necessary.